core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -95,7 +95,7 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
 
     if (hasChange) {
       return copyRecord(record, filteredFields);
-    } else if (filteredFields.size() == record.getFields().size()) {
+    } else if (record.getFields().size() != 0 && filteredFields.size() == record.getFields().size()) {
```

Comment:
This is the second bug mentioned, If you are pruning
`X : Struct { Y : Struct<>, Z: Int, ..... } `
And don't request "X" or "Y"

Y Has no fields so previously `filteredFields.size() = 0 = record.getFields().size()` which means we return
Y as a required field.

Reply:
This is what keeps "data_file" from being pruned on Unpartitioned tables (since partitionType would be an empty struct)

Reply:
I think we need a different operation that is the opposite of `GetProjectedIds`, like `ProjectFromIds`.

`TypeUtil.select` uses this class, `PruneColumns`, but it has behavior like a SQL `SELECT`. If I have a schema `a int, b struct<x double, y double>, c string` and I select `b`, then everything underneath `b` is selected, which is what you'd expect from `SELECT b FROM table`.

If we were to update `GetProjectedIds` with the logic above, then projecting `b struct<>` (which you can't do by naming columns) would actually result in the full struct getting projected because of the logic here that selects all of `b`. This class cannot be used to reconstruct a schema using the result of `GetProjectedIds`.

I think that we also need a `BuildProjection` that does the opposite of `GetProjectedIds` with the update to add empty structs. Then the datum reader could use that logic to prune the Avro schema and get an exact match with the expected schema.

Reply:
Oh yeah I forgot about that :/ yeah we'll need an inverse sort of thing.

Reply:
I tried to figure out if there was a simpler way to do this and I ended up modifying BuildAvroProjection. In my mind the real problem here was we were telling BuildAvroProjection what columns we wanted, but our pruned schema really only showed us what columns we needed that have data. This only left leaf empty-struct nodes. So during BuildAvroProjection we check for these empty-structs being requested in the expected schema and just add them back.

Reply:
I think there is a reasonable argument that `BuildAvroProjection` should do what your describe, since it is creating the final schema that will be requested from Avro.

If that's adding back the empty structs, then why change the behavior of `PruneColumns` here?

Reply:
This line is the change to fix bug 2,

**Example of Bug 2**
```
From (A, B, C, D {foo, bar, baz {}} ) // Starting with D containing a field baz which is empty
Project (A)
Returns
(A, D{ baz {} }) // Includes D even though it was not requested
```

The issue being that a struct which contains no elements naturally will always be included because
`records.filedsize == 0 == filteredFields.size == 0`

I think that behavior is incorrect, and this should be pruned out if it isn't needed in the projection

Reply:
Sorry, I meant to ask this on the block above.